pcf/particle/aws/kinesis/kinesis.py

```diff
@@ -0,0 +1,172 @@
+from pcf.util import pcf_util
+import logging
+import json
+import itertools
+import time
+from pcf.core import State
+from botocore.errorfactory import ClientError
+logger = logging.getLogger(__name__)
+
+class Kinesis(AWSResource):
+
+    flavor='kinesis'
+
+    #this is helpful if the particle doesn't have all three states or has more than three.
+
+    UNIQUE_KEYS = ["aws_resource.StreamName"]
+    equivalent_states = {
+            State.running: 1,
+            State.stopped: 0,
+            State.terminated: 0,
+        }
+
+    state_lookup = {
+        "CREATING":State.pending,
+        "DELETING" : State.pending,
+        "ACTIVE" : State.running,
+        "UPDATING" : State.running
+    }
+
+    START_PARAM_FILTER = {
+        "StreamName",
+        "ShardCount"
+    }
+    UPDATE_PARAM_FILTER = {
+        "StreamName",
+        "RetentionPeriodHours",
+        "EncryptionType",
+        "KeyId",
+        "Tags"
+    }
+    def __init__(self, particle_definition, session=None):
+        super().__init__(particle_definition, 'kinesis', session=session)
+        self.stream_name = particle_definition.get("aws_resource").get("StreamName")
+        self.get_desired_state_definition().get("Tags").append({"Key": "pcfName", "Value": particle_definition.get("pcf_name", '')})
+
+    def sync_state(self):
+        current_state = self.get_status()
+        if not current_state:
+            self.state = State.terminated
+            return
+        self.state = Kinesis.state_lookup.get(current_state.get("StreamStatus"))
+        self.current_state_definition = pcf_util.param_filter(current_state, Kinesis.UPDATE_PARAM_FILTER)
+        if not current_state.get("KeyId"):
+            self.current_state_definition["KeyId"] = ''
+        self.current_state_definition["Tags"] = self.client.list_tags_for_stream(StreamName=self.stream_name).get("Tags", [])
+        if current_state.get("StreamStatus") == "ACTIVE":
+            self.current_state_definition["ShardCount"] = self.getOpenShardCount()
+        else:
+            self.current_state_definition["ShardCount"] = 0
+
+    def getOpenShardCount(self):
+        current_state = self.get_status()
+        return current_state.get('OpenShardCount', 0)
+    def _update(self):
+        self._updateTags()
+        self._updateShards()
+        self._updateEncryptionTypeAndKey()
+        self._updateStreamRetention()
+
+    def _terminate(self):
+        return self.client.delete_stream(StreamName=self.stream_name)
+
+    def _start(self):
+        create_definition = pcf_util.param_filter(self.get_desired_state_definition(), Kinesis.START_PARAM_FILTER)
+        self.stream_name = self.get_desired_state_definition().get("StreamName")
+        if (self.get_status()):
+            self._update()
+        else:
+            self.client.create_stream(**create_definition)
+
+    def _stop(self):
+        return self._terminate()
+
+
+    def _updateShards(self):
+        current_state = self.get_status()
```

Comment:
No need to call ```current_state = self.get_status()``` multiple times. Just need to call it in sync_state

Reply:
I ran into an issue where I can't update multiple things at once and so if I base it all off of the sync_state get_status then I run into a possible issue where it attempts to update before finishing another update. 

For example, Let's say I have a kinesis stream that needs the shards to be updated and the encryption to be updated. 

On line 80 of kinesis.py updateShards() sees that the shards is off and updates the shards and puts in a update_shard_count call. The kinesis stream has now changed from 'ACTIVE' to 'UPDATING' however that change would not be reflected in the call to updateEncryptionTypeAndKey() and so it also sees that the encryption is different and tries to update the stream. This causes boto3 to throw an error because the stream is currently updating the shard count and thus can not update the encryption at this time. 

Thinking about it more, I could change it so that it catches the error though and just ignores it but I'm not sure what would be the best way to do this.

Reply:
never mind I think i figured out a good way to fix it.